src/game-logic/reducers/addExperience.test.js

```diff
@@ -0,0 +1,20 @@
+import { addExperience } from './addExperience'
+import { processLevelUp } from './processLevelUp'
+
+jest.mock('./processLevelUp')
+
+describe('addExperience', () => {
+  const gameState = { experience: 0, itemsSold: {} }
+
+  it('adds experience to current experience', () => {
+    const newState = addExperience(gameState, 100)
+
+    expect(newState.experience).toEqual(100)
+  })
+
+  it('will check for leveling up when experience is added', () => {
+    addExperience(gameState, 100)
+
+    expect(processLevelUp).toHaveBeenCalled()
```

Comment:
Ideally this unit test would not be concerned with implementation details like we're doing here. Could this test be changed to examine `gameState` to detect whether the level-up was processed instead?

Reply:
i have mixed feelings about this. not strong enough that I won't change it, but I do want to offer my perspective. i did things this way because I see this test as a unit test, and if it were to check that a level changed it would have to know about game logic conditions that cause the level to change (i.e. how much experience it takes to get from one level to another). the way i see, `addExperience` is only concerned with making sure that `processLevelUp` is invoked, which is what mocking it and checking for it to be called does. i think the place for doing a test on game logic belongs elsewhere, which is what might be missing from this PR, but i haven't found that spec yet 😄 .

Reply:
> the way i see, `addExperience` is only concerned with making sure that `processLevelUp` is invoked

I think this is where our views diverge. Based on the semantics and API of this reducer, I would assume that `addExperience` only cares about adding experience. The fact that level-ups are processed within this reducer seems like an implementation detail that could conceivably be handled outside of this function (though I'm not suggesting we do that).

We definitely do want to test that adding experience causes a level up, but perhaps this isn't the appropriate place to validate that? Specifically, I'm thinking that it may be better to validate that gaining experience results in a level-up in a higher-level React Test Library test. I'm thinking something similar to this: https://github.com/jeremyckahn/farmhand/blob/f791e83c4d4994e30efbdaddead4c446118f598c/src/game-logic/item-sell.test.js#L9-L29

I realize that this is expanding the scope of what you set out to do here, and the current state isn't something I'd block this PR over or anything. It's just a difference of opinion. If you feel good about what we have here, then I'm okay with sticking with it. Just let me know which way you'd like to go. 🙂

Reply:
i actually think i'm with you on processing level ups as being the thing that is out of place here. so i think i'll remove that. it definitely feels like an unexpected side effect not that you've mentioned it

Reply:
ok i've removed it. i see that `sellItem` is what actually does `processLevelUp` but I don't see any tests in there explicitly checking for level up. i always forget that `level` isn't part of farmhand state and that it's calculated when needed so testing that a level up was achieved is trickier than i expected (at least, without just mocking and checking for the call.) if you have suggestions on how to test it let me know and i am happy to add it to this PR

Reply:
Thanks for making that change!

> testing that a level up was achieved is trickier than i expected (at least, without just mocking and checking for the call.)

If we're speaking in terms of React Testing Library tests, I'd approach this by checking to see if the level up notification text is present on the page: https://github.com/jeremyckahn/farmhand/blob/d37d41073867da6a8dbfa65048c0b008b303a9f5/src/templates.js#L184-L214

Reply:
ahh yes that makes sense. the `sellItem` test doesn't seem like the right place in that case (it isn't doing any rendering). i suppose we could check the notifications in state from there though?